### ISSUE TITLE
[BUG]Tuned_model does not return most optimized model

### ISSUE THREAD

**Issue by dandluu** [pycaret/pycaret#1987]
**Describe the bug**

Unsure if this is a bug or if I am doing something wrong, I can change the label of this report.
I created a tuned model following the basic workflow from pycaret and the score grid gave me a variety of results from its score grid. 
I expected that the model returned is the most 'optimized' model however that did not seem to be the case. From plotting the model errors the model still shows a low r2 score. Additionally, when using that model to predict on the already trained X_test_data, I still get poorer results than from what was shown in the grid score for the model. 

**To Reproduce**

Due to the sensitive nature of the data, I am not able to provide data source, but I can continually reproduced the error from my dataset by following pycaret's basic workflow to generate a random forest regressor on a tuned model. 

```python

model_type = 'rf'
n_iter=50
optimize="r2"
df = df[column_selection]
start_up = setup(data = df[column_selection], target = target, train_size=0.8, silent=True, preprocess=False)   

model = create_model(model_type)
tuned_model = tune_model(model, n_iter=n_iter, optimize=optimize, choose_better=True)

plot_model(tuned_model, plot='error')

```

**Expected behavior**

I expected the model to return the model parameters that yields the best r2 score and its hyper parameters. See details below. The model should have return the hyper parameters with that gave the highest r2 which was the selected feature to optimize.

Is there anyway to capture that from the output grid-score table, like what were the hyper-parameters that generated that score?

**Additional context**

**Notebook model creation**
![model_creation](https://user-images.githubusercontent.com/26827591/147142622-52f71521-4b56-408b-b4bf-f9c628a7509e.PNG)

**tuned_model Output Results**
![sample_results_25](https://user-images.githubusercontent.com/26827591/147142628-9417c64d-78df-4bbd-b42e-170465e7ada6.PNG)
**Continued Results with high scores being at the bottom, next cell sort results by R2**
![sample_results_last_sort_97](https://user-images.githubusercontent.com/26827591/147142630-33d993b0-caf0-46d0-842d-ce02b46e40ff.PNG)

**Model Plot**
![plot_model_r2_27](https://user-images.githubusercontent.com/26827591/147143922-daa62291-413c-438b-9d1c-587680de3404.PNG)

**Versions**
<details>

pycaret version == '2.3.5'
</details>

**Comment by moezali1** [comment]
@dandluu Is it possible for you to reproduce this bug using any open-source public dataset?

**Comment by dandluu** [comment]
Hi @moezali1, I was able to use a different dataset and get a similar issue, however with this dataset, the returned model is sometimes optimized, other times it is not. In the case of this issue, I have reproduced the bug and provided the notebook as well. In this case, the returned model has a fit of ~61% however, in training it produced a model with an r2 of ~84% 
[notebook with bug](https://github.com/dandluu/pycaret_sol_bug/blob/main/solubility_pycaret_bug.ipynb)

**Comment by Frank3W** [comment]
Based on the notebook shared, I think the data frame returned by pull() is the cross validation metric results for the model with best hyper parameters (after tuning process). By default, there are 10 folds. That's why 10 results shown in that table. n_iter=20 is used in hyper parameter tuning process to control the number of iterations. It is not on the number of folds. The mismatch between R2 61% and 84%, I guess, is because the function plot_model shows R2 on test data but the data frame from pull() shows R2 on each fold.

**Comment by dandluu** [comment]
Thank you for clarifying the n_iter and folds. However, how does the optimize model work, does the model store the hyper parameters from the cv metric results that gave the highest result during its kfold cv? Is there a way to tease out and store the train/test for each fold during training so we get the train/test that resulted in the highest r2 during training? I say this because the data is complex so by splitting the folds we are looking further into the similarities between the datasets.

**Comment by moezali1** [comment]
@dandluu The model returned by `tune_model` function is the model that has best hyperparameters. The results you see on screen or you access using `pull` function is the cross-validated results of best selected hyperparameters. If you need access to trials generated as a result of `n_iter` then you can pass `return_tuner=True` in the `tune_model` function. 

More details on this link: https://pycaret.gitbook.io/docs/get-started/functions/optimize#access-the-tuner
